### PR TITLE
Refactor animations to better allow animations of color and other components.

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description
Changes the Animation interface to provide an abstraction that allows components to define an  "animation space vector". This vector needs to work like a cartesian coordinate and n-dimensional space and provide a way to be constructed from the given component (and construct a component from a given a nimation space vector.

For basic datatypes like `f64` `Vector4` the value itself is identical to the animation space vector, but e.g., for `color::Rgba` we want animations to happen in an aproriate color space and thus the animation space vector is a `Vector4` in the `color::Laba` space.

### Important Notes
Precondition for merging  #834 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has been tested where possible.
- [x] All code has been profiled where possible.

### CODE REVIEW

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 1]
```diff
@@ -189,31 +189,91 @@ impl<T> display::Object for ShapeView<T> {
...
 /// Smart animation handler. Contains of dynamic simulation and frp endpoint. Whenever a new value
 /// is computed, it is emitted via the endpoint.
-#[derive(CloneRef,Derivative,Debug,Shrinkwrap)]
+#[derive(CloneRef,Derivative,Debug)]
 #[derivative(Clone(bound=""))]
 #[allow(missing_docs)]
-pub struct Animation<T> {
-    #[shrinkwrap(main_field)]
-    pub simulator : DynSimulator<T>,
+pub struct Animation<T:Animatable<T>> {
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:239]: Oh, this looks bad! `T:Animatable<T>`. You should never have traits which refer to itself this way. You defintely could have written it as `T:Animatable` instead - see my impl for reference

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 2]
```diff
@@ -189,31 +189,91 @@ impl<T> display::Object for ShapeView<T> {
...
 // =================
 
 /// Smart animation handler. Contains of dynamic simulation and frp endpoint. Whenever a new value
 /// is computed, it is emitted via the endpoint.
-#[derive(CloneRef,Derivative,Debug,Shrinkwrap)]
+#[derive(CloneRef,Derivative,Debug)]
 #[derivative(Clone(bound=""))]
 #[allow(missing_docs)]
-pub struct Animation<T> {
-    #[shrinkwrap(main_field)]
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:202]: why this change? It makes usage ugly

File: src/rust/ensogl/lib/core/src/gui/cursor.rs [hunk 3]
```diff
@@ -367,8 +367,8 @@ impl Cursor {
         radius.set_target_value(DEFAULT_RADIUS);
         size.set_target_value(DEFAULT_SIZE());
 
-        let fade_out_spring = inactive_fade.spring() * 0.2;
-        let fade_in_spring  = inactive_fade.spring();
+        let fade_out_spring = inactive_fade.simulator.spring() * 0.2;
+        let fade_in_spring  = inactive_fade.simulator.spring();
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/cursor.rs:371]: forcing user to know that "animation" uses "simulator" under the hood is leaking internal design to the api. its bad

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: src/rust/ensogl/lib/core/src/data/color.rs [hunk 4]
```diff
@@ -19,6 +19,7 @@ pub mod component;
 pub mod data;
 pub mod gradient;
 pub mod space;
+pub mod animations;
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/data/color.rs:22]: Moreover, this is not responsibility of colors to think about animations, so this modul;e placemetn is bad. But read my other comments - they propose a better design which actually suits this place

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 5]
```diff
@@ -189,6 +189,61 @@ impl<T> display::Object for ShapeView<T> {
...
+
+/// Shorthand for a  HasAnimationSpaceRepr::AnimationSpaceRepr wrapped in a `AnimationLinearSpace`.
+pub type AnimationSpaceRepr<T> = AnimationLinearSpace<<T as HasAnimationSpaceRepr>::AnimationSpaceRepr>;
+
+/// This is used to make type inference better on usage places - without it,
+/// Rust would force one to write the where part EVERYWHERE in use places.
+pub trait FromAsInto<T> = Sized where T:Into<Self>;
+
+/// Can be transformed from and into.
+pub trait FromInto<T> = Sized + Into<T> + FromAsInto<T>;
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:218]: Why not `Iso`? This is a standard form of isomorphism and imo this sohuld be named this way. Moreover, why not placing it in prelude like I proposed in my snippet?

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 6]
```diff
@@ -198,22 +253,45 @@ impl<T> display::Object for ShapeView<T> {
...
+pub struct Animation<T:Animatable> {
     #[shrinkwrap(main_field)]
-    pub simulator : DynSimulator<T>,
+    pub simulator : DynSimulator<T::AnimationSpaceRepr>,
     pub value     : frp::Stream<T>,
 }
 
-impl<T:inertia::Value> Animation<T> {
+#[allow(missing_docs)]
+impl<T:Animatable+Debug+Default+Clone+'static> Animation<T> {
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:263]: Are these constraints all needed? I dont get where `'static'`, `Default`, and `Debug` comes from.

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 7]
```diff
@@ -198,22 +253,45 @@ impl<T> display::Object for ShapeView<T> {
...
     /// Constructor.
     pub fn new(network:&frp::Network) -> Self {
         frp::extend! { network
             def target = source::<T>();
         }
-        let simulator = DynSimulator::<T>::new(Box::new(f!((t) target.emit(t))));
+        let simulator = DynSimulator::<T::AnimationSpaceRepr>::new(Box::new(f!((t) {
+             target.emit(AnimationLinearSpace{value:t}.into())
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:270]: Change it pls to `t.to_animation_linear_space()` - this can be a method generated automatically which uses `into` under the hood. Lets apply it to other places below as well :) Manual construction of wrappers everywhere adds a lot of boilerplate

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 5 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/lib/core/src/gui/component.rs:218]: We don't really define a mathematical isomorphism here. That is [a lot more specific](https://en.wikipedia.org/wiki/Isomorphism). All we do require is a `From<T>` and `Into<T>` implementation. From a Rust perspective, the name `FromInto` is much clearer about the requirements. 

I agree, needs to move to prelude though.

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 6 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/lib/core/src/gui/component.rs:263]: I think they come from the FRP. I will check that later.

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 7 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/lib/core/src/gui/component.rs]: I added a `from_animation_space` function that does that.

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 5 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/lib/core/src/gui/component.rs:218]: > I agree, needs to move to prelude though.

Just noticed prelude is in a different crate, though I would leave this here for now. If we feel strongly about the usefulness, we can open an issue to add it there and refactor here after it has landed.

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 6 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/lib/core/src/gui/component.rs:268]: Yeah, all of the contsraints are needed for the FRP. Otherwise `T` can't be used in `source::<T>();`

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 5 above)
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/gui/component.rs:218]: The fact that prelude is in different crate should not prevent us from changing it. This should definitely be in prelude IMO. If youre not sure, lets talk about it. Also, lets talk more aobut the Iso name, as I believe this would be the correct name for that.

File: src/rust/ensogl/lib/core/src/gui/component.rs (hunk 6 above)
Comment (wdanilo) [reply src/rust/ensogl/lib/core/src/gui/component.rs:268]: If they come from FRP, please use their FRP type alias so we know they are from FRP! Expanding constraitns like this make it error prone in the future - they will change in frp and we will not know where thay came from. Instead, always try to use aliases prepared for it. So just use `T:frp::Data` instead!

File: src/rust/ensogl/lib/core/src/gui/component.rs [hunk 8]
```diff
@@ -198,22 +251,45 @@ impl<T> display::Object for ShapeView<T> {
...
+        from_animation_space(value)
+    }
+
+    pub fn set_target_value(&self, target_value:T) {
+        let state:AnimationLinearSpace<_> = target_value.into();
+        self.simulator.set_target_value(state.value);
+    }
+
+    pub fn target_value(&self) -> T {
+        let value =  self.simulator.target_value();
```
Comment (wdanilo) [inline src/rust/ensogl/lib/core/src/gui/component.rs:290]: double space

**Review by wdanilo — APPROVED** [review]
